pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -222,6 +242,38 @@ func (f *framework) RunPrefilterPlugins(
 	return nil
 }
 
+// RunScorePlugins runs the set of configured scoring plugins. It returns a map that
+// stores for each scoring plugin name the corresponding NodeScoreList(s).
+// It also returns *Status, which is set to non-success if any of the plugins returns
+// a non-success status.
+func (f *framework) RunScorePlugins(pc *PluginContext, pod *v1.Pod, nodes []*v1.Node) (PluginToNodeScoreMap, *Status) {
+	pluginToNodeScoreMap := make(PluginToNodeScoreMap, len(f.scorePlugins))
+	for _, pl := range f.scorePlugins {
+		pluginToNodeScoreMap[pl.Name()] = make(NodeScoreList, len(nodes))
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan interface{}, 1)
+	workqueue.ParallelizeUntil(ctx, 16, len(nodes), func(index int) {
+		for _, pl := range f.scorePlugins {
+			score, status := pl.Score(pc, pod, nodes[index].Name)
+			if !status.IsSuccess() {
+				schedulerutil.NonBlockingSend(errCh, status.Message())
+				cancel()
+				return
+			}
+			pluginToNodeScoreMap[pl.Name()][index] = score * int(f.pluginNameToWeightMap[pl.Name()])
```

Comment:
There is a chance that the name the plugin returns does not exist in the `pluginNameToWeightMap`.

Another point, why are returning scores per plugin per node? This requires a lot more memory. We could just add them here and return one score per node. Besides, writing to a map concurrently requires synchronization. However, writing concurrently to disjoint elements of a pre-allocated slice does not need synchronization. So, if we return a slice of node scores, we can build that slice here without paying the overhead of synchronization.

Reply:
> There is a chance that the name the plugin returns does not exist in the `pluginNameToWeightMap`.

Done

> Another point, why are returning scores per plugin per node? This requires a lot more memory. We could just add them here and return one score per node. 

I was assuming that the NormalizeScore plugins expect the nodes' scores of its corresponding Score plugin. That is what the map-reduce priority functions currently work.

> Besides, writing to a map concurrently requires synchronization. However, writing concurrently to disjoint elements of a pre-allocated slice does not need synchronization. So, if we return a slice of node scores, we can build that slice here without paying the overhead of synchronization.

This does not write to the map, the slice is pre-allocated per plugin already (see the beginning of the function), each iteration of this parallel loop writes to an index in the pre-allocated slice.

Reply:
By the way, currently we [apply the weight](https://github.com/kubernetes/kubernetes/blob/d59a603f1b32793a4e7a9681f620344101d24f3e/pkg/scheduler/core/generic_scheduler.go#L771) for the priority functions after the reduce, and so in our case it will be after NormalizeScore. I am thinking we may actually need to do the same: basically after RunNormalizeScore, we run a function WeightScores that just applies the weight.

Reply:
> Another solution is to run Score, NormalizeScore and apply the weight here in this function, which also allow us to avoid returning the multi-dimensional map.

This sounds good to me, but please do it in a separate PR for normalize scoring.